### PR TITLE
cli: snapcraft promote

### PR DESCRIPTION
**Pull request by sergiusens**
Create a snap Status/Status class to easily handle the state response
and construct a channel set to use with the promote command.

LP: #1827513
Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

**Comment by cprov** [comment]
Thanks @sergiusens, the implementation looks good and easy to update to the actual promote endpoint when it exists in the store. Meanwhile, there is a remote possibility of revisions being partially promoted (if an attempt fails), running the same promotion command will fix it.

File: snapcraft/storeapi/status.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    def __repr__(self) -> str:
+        return "<SnapStatusChannelDetails: {} on {}>".format(
+            self._snap_name, self._arch
+        )
+
+    def __iter__(self):
+        for attr in ["arch", "revision", "version"]:
+            yield self.__getattribute__(attr)
+
+    def __init__(self, *, snap_name: str, arch: str, payload) -> None:
```
Comment (cmatsuoka) [inline snapcraft/storeapi/status.py]: Could you add a type annotation for `payload`?

File: snapcraft/storeapi/status.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    @property
+    def revision(self) -> Optional[int]:
+        return self._payload.get("revision")
+
+
+class SnapStatusTrackDetails:
+    def __repr__(self) -> str:
+        return "<SnapStatusTrackDetails: {} on {}>".format(self._snap_name, self.track)
+
+    def __init__(self, *, snap_name: str, track: str, payload) -> None:
```
Comment (cmatsuoka) [inline snapcraft/storeapi/status.py]: Could you add a type annotation for `payload`?

**Comment by sergiusens** [comment]
Thanks @cprov wrt promotion being individual releases, we do indeed suffer from this possibily, just like on the web interface, although it is easier to surface on a GUI than on a CLI

Thanks for the compliments @cmatsuoka, I will get to the type annotations.

File: snapcraft/cli/store.py [hunk 3]
```diff
@@ -226,6 +236,92 @@ def release(snap_name, revision, channels):
...
+    if parsed_from_channel == parsed_to_channel:
+        raise click.BadOptionUsage(
+            "--from-channel and --to-channel cannot be the same."
+        )
+    elif parsed_from_channel.risk == "edge" and parsed_from_channel.branch is None:
+        raise click.BadOptionUsage(
+            "{!r} is not a valid set value for --from-channel.".format(
+                parsed_from_channel
+            )
+        )
```
Comment (kyrofa) [inline snapcraft/cli/store.py:296]: What's up with this limitation? Bites me all the time.

**Comment by sergiusens** [comment]
I had a conversation and it is going away

**Pull request merged**